Src/ILGPU/Runtime/ExchangeBuffer.cs

```diff
@@ -381,4 +458,682 @@ protected override void Dispose(bool disposing)
+
+        /// <summary>
+        /// Represents a view source that allocates native memory in page-locked CPU
+        /// memory.
+        /// </summary>
+        sealed class CudaViewSource : ViewPointerWrapper
+        {
+            /// <summary>
+            /// Creates a new Cuda view source.
+            /// </summary>
+            /// <param name="sizeInBytes">The size in bytes to allocate.</param>
+            /// <returns>An unsafe array view source.</returns>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static CudaViewSource Create(int sizeInBytes)
+            {
+                CudaException.ThrowIfFailed(
+                    CudaAPI.Current.AllocateHostMemory(
+                        out IntPtr hostPtr,
+                        new IntPtr(sizeInBytes)));
+                return new CudaViewSource(hostPtr);
+            }
+
+            /// <summary>
+            /// Creates a new unmanaged memory view source.
+            /// </summary>
+            /// <param name="nativePtr">The native host pointer.</param>
+            private CudaViewSource(IntPtr nativePtr)
+                : base(nativePtr)
+            { }
+
+            /// <summary cref="ArrayViewSource.GetAsRawArray(
+            /// AcceleratorStream, Index1, Index1)"/>
+            protected internal override ArraySegment<byte> GetAsRawArray(
+                AcceleratorStream stream,
+                Index1 byteOffset,
+                Index1 byteExtent) => throw new InvalidOperationException();
+
+            #region IDispoable
+
+            /// <summary cref="DisposeBase.Dispose(bool)"/>
+            protected override void Dispose(bool disposing)
+            {
+                if (NativePtr != IntPtr.Zero)
+                {
+                    CudaAPI.Current.FreeHostMemory(NativePtr);
+                    NativePtr = IntPtr.Zero;
+                }
+                base.Dispose(disposing);
+            }
+
+            #endregion
+        }
+
+        #endregion
+
+        #region Instance
+
+        /// <summary>
+        /// The internally allocated CPU memory.
+        /// </summary>
+        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+        private readonly ViewPointerWrapper cpuMemory;
+
+        /// <summary>
+        /// A cached version of the CPU memory pointer.
+        /// </summary>
+        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+        private readonly void* cpuMemoryPointer;
+
+        /// <summary>
+        /// Initializes this memory buffer.
+        /// </summary>
+        /// <param name="buffer">The underlying memory buffer.</param>
+        /// <param name="mode">The current buffer allocation mode.</param>
+        internal ExchangeBuffer2D(MemoryBuffer<T, Index2> buffer, ExchangeBufferMode mode)
+            : base(buffer.Accelerator, buffer.Extent.Size)
+        {
+            // Allocate CPU memory
+            cpuMemory = Accelerator is CudaAccelerator &&
```

Comment:
Same here: please try to remove the amount of duplicated code as much as possible to reduce maintenance and bug-fixing overhead